brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,169 @@
+      LOG(ERROR) << "get fd flags";
+    flags |= FD_CLOEXEC;
+    if (fcntl(pipehack[i], F_SETFD, flags) == -1)
+      LOG(ERROR) << "set fd flags";
+  }
+
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIGCHLDHandler;
+  sigaction(SIGCHLD, &action, NULL);
+}
+#endif
+
+namespace brave {
+
+class TorLauncherDelegate : public base::LaunchOptions::PreExecDelegate {
+ public:
+    TorLauncherDelegate() {}
+    ~TorLauncherDelegate() override {}
+
+    void RunAsyncSafe() override {
+#if defined(OS_POSIX)
+      for (size_t i = 0; i < 2; ++i)
+        close(pipehack[i]);
+      setsid();
+#endif
+    }
+ private:
+    DISALLOW_COPY_AND_ASSIGN(TorLauncherDelegate);
+};
+
+TorLauncherImpl::TorLauncherImpl(
+    std::unique_ptr<service_manager::ServiceContextRef> service_ref)
+    : service_ref_(std::move(service_ref)) {
+  SetupPipeHack();
+}
+
+TorLauncherImpl::~TorLauncherImpl() {
+  if (tor_process_.IsValid()) {
+    tor_process_.Terminate(0, false);
+#if defined(OS_POSIX)
+    for (size_t i = 0; i < 2; ++i)
+      close(pipehack[i]);
+#endif
+#if defined(OS_MACOSX)
+    base::PostTaskWithTraits(
+        FROM_HERE, {base::MayBlock(), base::TaskPriority::BACKGROUND},
+        base::BindOnce(&base::EnsureProcessTerminated, Passed(&tor_process_)));
+#else
+    base::EnsureProcessTerminated(std::move(tor_process_));
+#endif
+  }
+}
+
+void TorLauncherImpl::Launch(const base::FilePath& tor_bin,
+                             const std::vector<std::string>& args,
+                             LaunchCallback callback) {
+  base::CommandLine cmdline(tor_bin);
+  for (size_t i = 0; i < args.size(); ++i) {
+    cmdline.AppendArg(args[i]);
+  }
+  base::LaunchOptions launchopts;
+  TorLauncherDelegate tor_launcher_delegate;
+  launchopts.pre_exec_delegate = &tor_launcher_delegate;
+#if defined(OS_LINUX)
+  launchopts.kill_on_parent_death = true;
+#endif
+  tor_process_ = base::LaunchProcess(cmdline, launchopts);
+
+  bool result;
+  // TODO(darkdh): return success when tor connected to tor network
+  if (tor_process_.IsValid())
+    result = true;
+  else
+    result = false;
+
+  if (callback)
+    std::move(callback).Run(result);
+
+  if (!child_monitor_thread_.get()) {
+    child_monitor_thread_.reset(new base::Thread("child_monitor_thread"));
+    if (!child_monitor_thread_->Start()) {
+      NOTREACHED();
+    }
+
+    child_monitor_thread_->task_runner()->PostTask(
+      FROM_HERE,
+      base::BindOnce(&TorLauncherImpl::MonitorChild, base::Unretained(this)));
+  }
+}
+
+void TorLauncherImpl::SetCrashHandler(SetCrashHandlerCallback callback) {
+  crash_handler_callback_ = std::move(callback);
+}
+
+void TorLauncherImpl::MonitorChild() {
+#if defined(OS_POSIX)
+  char buf[PIPE_BUF];
+
+  while (1) {
+      if (read(pipehack[0], buf, sizeof(buf)) > 0) {
+        pid_t pid;
+        int status;
+
+        if ((pid = waitpid(-1, &status, WNOHANG | WUNTRACED)) != -1) {
```

Comment:
Using WUNTRACED here causes us to be notified if tor is suspended with SIGSTOP.  It may be resumed later with SIGCONT, but the following logic assumes this means it has terminated.  Is there a reason we need to use WUNTRACED here that's not obvious to me?

Also: The paranoiac in me would double-check that pid here is the pid we expect of the child; if not, report a confused warning message and ignore the event.

Reply:
`WUNTRACED` is intentional if we don't use it, `SIGSTOP` will fall into `WIFSIGNALED` instead of `WIFSTOPPED`.

Reply:
SIGSTOP doesn't fall into WIFSIGNALED because WIFSIGNALED means the program _terminated_ on a signal, which SIGSTOP doesn't do.

You can see with the enclosed test program: `./test cat` runs `cat` and waits for it without WUNTRACED, _i.e._ ignoring the stop signals, whereas `./test -s cat` runs `cat` and waits for it with WUNTRACED; you can see what happens when you `kill -STOP <pid>` and `kill -TERM <pid>` and `kill -CONT <pid>` in another terminal.

[test.c.txt](https://github.com/brave/muon/files/1890956/test.c.txt)

Reply:
hmm. there must be some weird things happened when I handle waitpid in SIGCHLD handler.
I will remove this.

Reply:
fixed in 4e72ec6